### ISSUE TITLE
Derive a form (multiple prompts) from struct

### ISSUE THREAD

**Issue by simonsan** [ynqa/promkit#11]
I would like to easily annotate a configuration struct, to ask the user for input for various fields and give a hint on validation:

```rust

use derive_bla::AskUser;

#[derive(Debug, AskUser)]
struct MyConf {
    #[ask(q="Please enter password:", hide_input=true, min_len=12)]
    password: Option<String>,

    #[ask(q="Please enter username:", min_len=3)]
    username: Option<String>,

    #[ask(q="Please enter age:", range=18..=85)]
    age: Option<u8>,
}

fn main() -> Result<()> {
    let mut conf = MyConf::default()
    conf.password()?.validate()?;
    conf.username()?.validate()?;
    conf.age()?.validate()?;

    println!("{conf#?}");
}
```

I would be interested if something like this is already existing, e.g. based on inquire, promkit, or if there is any interest to add this?

There are some ideas already in `inquire`, but it feels a bit out of scope.

CC: https://github.com/mikaelmello/inquire/issues/212
CC: https://github.com/mikaelmello/inquire/issues/65

EDIT: It would basically be the `clap` way of doing interactive prompts. Annotate something and give it to the user to let it be filled out. I think it comes with its own advantages and disadvantages, and excels a lot in a workflow where you suddenly want the user to fill out some data you already have bundled in a struct. 

I think it could be a game changing feature, though. Because it would make dealing with interactive prompts much easier for a lot of users. Also, it would be excellent to have for people that don't want the full-fledged control of a `ratatui` TUI, but still a bit more ergonomic when it comes to interactivity and developer experience creating such.

**Comment by ynqa** [comment]
@simonsan Thank you for your proposal!

Currently, we do not offer such a feature (this is simply because I am not yet well-versed in Derive/macro). However, I'm interested in it.

**Comment by simonsan** [comment]
Nice! In one issue, there is an example/demo of how such a thing could work: https://github.com/IniterWorker/inquire_derive

Maybe it helps to get started. Also, there is the [The Little Book of Rust Macros](https://veykril.github.io/tlborm/proc-macros/methodical/derive.html) and there is the [proc-macro workshop by dtolnay](https://github.com/dtolnay/proc-macro-workshop). I'm saying this, because I'm trying to support and hope that such a thing will exist at one point in time, as I think it would be really valuable to the ecosystem. (:

**Comment by ynqa** [comment]
@simonsan I've started by creating a prototype. I believe it needs to be refined further, but what do you think?
- https://github.com/ynqa/promkit/blob/dev-0.4.0/promkit-derive/promkit-derive/examples/example.rs

**Comment by simonsan** [comment]
Looks really nice! 👍🏽 I need to try it our, to see if the API is ergonomic for me. But the way you instantiate and do it, is exactly how I would imagine it.

- returning a result, so we can return early with the try operator after each step

- `validate` could be implemented on top of it at later stage, there is a [`validator`](https://crates.io/crates/validator) crate, which API is pleasant and could be inspiring. T.validate_args((77, 555)).is_ok()` sound like a good pattern to me as well

Great work! 🌷

**Comment by ynqa** [comment]
@simonsan Thank you for your comment. Please feel free to continue post your feedback after you have had a chance to try it out.  As a favor, could I ask for the opportunity to have you take another look at the macro after I've refined it based on your feedback?

**Comment by ynqa** [comment]
@simonsan Thank you for your comment! 👍 Please feel free to continue post your feedback after you have had a chance to try it out. As a favor, could I ask for the opportunity to have you take another look at the macro after I've refined it based on your feedback?

**Comment by simonsan** [comment]
> @simonsan Thank you for your comment! 👍 Please feel free to continue post your feedback after you have had a chance to try it out. As a favor, could I ask for the opportunity to have you take another look at the macro after I've refined it based on your feedback?

For sure! :) 👍🏽

**Comment by ynqa** [comment]
@simonsan I refined it in PR #17. Here are the changes:

- Changed the term ask to readline. This indicates the use of the preset [readline](https://github.com/ynqa/promkit/blob/main/src/preset/readline.rs). Any parameters that can be set via the macro attribute will directly apply as parameters for readline.
- Added #[readline(default)]. This defines readline as the default.
- The supported field types are Option<T> and T (where T satisfies `FromStr`).

I understand that changing the name from ask might raise concerns or there may still be issues to consider. Please share your thoughts.

Example:

```rust
use promkit::{crossterm::style::Color, style::StyleBuilder, Result};
use promkit_derive::Promkit;

#[derive(Default, Debug, Promkit)]
struct Profile {
    #[readline(
        prefix = "What is your name?",
        prefix_style = StyleBuilder::new().fgc(Color::DarkCyan).build(),
    )]
    name: String,

    #[readline(default)]
    hobby: Option<String>,

    #[readline(prefix = "How old are you?", ignore_invalid_attr = "nothing")]
    age: usize,
}

fn main() -> Result {
    let mut ret = Profile::default();
    ret.readline_name()?;
    ret.readline_hobby()?;
    ret.readline_age()?;
    dbg!(ret);
    Ok(())
}
```

**Comment by simonsan** [comment]
I'm on the way to going out, so only tiny feedback:

- I don't mind, if it's called `readline`, or ask, or something. I think it should make clear the intent, of what is happening, and `readline` might be sufficient for that. Could also be `input` for example.

- `prefix` I personally find a bit confusing, I would rather call it `prompt` as this would make it clear what it will do

- for the supported fields I think something like a `Vec` and a `HashSet` (although we can build that from a `Vec` might be good as well, like clap does it:

```rust
    /// Tags for the activity
    #[cfg_attr(
        feature = "clap",
        clap(
            short,
            long,
            group = "adjust",
            name = "Tags",
            value_name = "Tags",
            visible_alias = "tag",
            value_delimiter = ','
        )
    )]
    tags: Option<Vec<String>>,
```

**Comment by simonsan** [comment]
For the `readline` attribute:
- `user_prompt`
- `input_prompt`
- `question`
- `user_input`

For the `prefix` attribute:
- `prompt_text`
- `question_text`
- `message`
- `query`

**Comment by ynqa** [comment]
@simonsan Thanks for your comments! 👍 

Indeed, the terminology needs to be clearer about what can be done. It felt like it could also be offered as something separate from presets.

Accepting collections with delimiters is a good idea. On the other hand, incorporating various parser logics might make the macro complex (it already seemed quite complex when allowing for `Option`). Therefore, what do you think about accepting parser functions in the attribute for types beyond primitives, like `str -> Vec<T>`, `str -> Option<T>` (I believe this was also in clap)?

**Comment by simonsan** [comment]
A form could be even layouted like this:

```console
Please fill out the form:
------------------------
What is your name?: John Doe
What is your hobby? (Leave empty if none):
How old are you?: 25

Use Up/Down to navigate through the inputs
Press Enter if you are ready to continue
```

I don't know if that is easily possible, though 😅 It's not a dealbreaker, would just double down on the `Form`

**Comment by simonsan** [comment]
> Accepting collections with delimiters is a good idea. On the other hand, incorporating various parser logics might make the macro complex (it already seemed quite complex when allowing for `Option`). Therefore, what do you think about accepting parser functions in the attribute for types beyond primitives, like `str -> Vec<T>`, `str -> Option<T>` (I believe this was also in clap)?

I agree completely. We don't want to make it to complex. I think `clap` has https://docs.rs/clap/latest/clap/builder/struct.ValueParser.html for this.

**Comment by simonsan** [comment]
I found this one from the Go ecosystem, which might be also interesting from a design perspective:
https://github.com/charmbracelet/huh

**Comment by ynqa** [comment]
> When you are feeling ready for a review, let me know, I'll be able to look over it then. 🫂

From [this comment](https://github.com/ynqa/promkit/pull/17#issuecomment-2429343468).

@simonsan I was on a bit of a journey to working on some projects like [jnv](https://github.com/ynqa/jnv) and [sig](https://github.com/ynqa/sig) (and also doing the actual travel). I'm gradually returning to this development. Apologies for the delay in my response.

> A form could be even layouted like this:

I’ve prepared the form-like UI ([ref](https://github.com/ynqa/promkit?tab=readme-ov-file#form)) mentioned in [your previous comment](https://github.com/ynqa/promkit/issues/11#issuecomment-2007271666) and am currently updating the derive to use this form instead of `readline`. Once that is complete, I'll formally submit a pull request.
